### PR TITLE
Fix bug, Parse github event properly & improvements

### PR DESCRIPTION
**Pull request by hauvong**
* Listen channel message with Github bot event that closed the pull request.
* Permanent delete when remove action is requested.
* Add Close button to app home page listing.
* Filter listing by channel on app home page.
* Add user session to keep track of each user in app home page.
* Add setting for repository number of review & approval required
* Add setting to set who to mention per channel when new code review is requested
* Fix bug where change request may not update request status properly
* Improve Web UI with support of slack channel filter.
  Screenshot from 2023-02-27 09-16-51
 Screenshot from 2023-02-27 11-47-28

### CODE REVIEW

File: client/src/lib/RenderReviewList.js
Comment (RomanGluschukDataart) [inline client/src/lib/RenderReviewList.js:49]: Look like you can reuse handlePageChange method here

File: client/src/lib/RenderReviewList.js
Comment (RomanGluschukDataart) [inline client/src/lib/RenderReviewList.js:39]: fetchData( { ...
can be defined as separate method and called from 39, 51, 66 , 104 lines

**Review by RomanGluschukDataart — CHANGES_REQUESTED** [review]

File: client/src/lib/RenderReviewList.js
Comment (hauvong) [reply client/src/lib/RenderReviewList.js:49]: Not really, because we do not want to trigger the unneeded side affects against other data input states

File: client/src/lib/RenderReviewList.js
Comment (hauvong) [reply client/src/lib/RenderReviewList.js:39]: The fetchData call signature was implement this way to allow finer control on how each data is fetch base on their respective data change after and before the side affects trigger.  Prior version had a combine reused side affects variables, prove to be harder to troubleshoot and improve when addition features are added because each state cause by side affects will cause data to refresh and depending on how reach and ordered of evens fired, the incorrect data is called by fetchData.

**Review by RomanGluschukDataart — APPROVED** [review]

File: client/src/lib/RenderReviewList.js
Comment (RomanGluschukDataart) [reply client/src/lib/RenderReviewList.js:39]: Thank, you for your wide response.

File: client/src/lib/RenderReviewList.js
Comment (RomanGluschukDataart) [reply client/src/lib/RenderReviewList.js:49]: Ok, fine.

File: src/bolt/utils.ts
Comment (pattiereaves) [inline src/bolt/utils.ts:238]: Should this use straight quotes since they aren't dynamic?

File: src/lib/config.ts
Comment (pattiereaves) [inline src/lib/config.ts:92]: this function name would be more readable as `setRepositoryNumberOfReviews`

File: src/lib/config.ts
Comment (pattiereaves) [inline src/lib/config.ts:104]: Likewise `setRepositoryNumberOfApprovals`

File: src/lib/config.ts
Comment (pattiereaves) [inline src/lib/config.ts:66]: this function name would be more readable as `getRepositoryNumberOfReviews`

File: src/lib/config.ts
Comment (pattiereaves) [inline src/lib/config.ts:79]: `getRepositoryNumberOfApprovals`

**Review by pattiereaves — DISMISSED** [review]
with a few suggestions, can approve again if they are applied.

File: client/src/lib/RenderReviewList.js
Comment (hauvong) [reply client/src/lib/RenderReviewList.js]: Have try that, also tried useRef.  Because the state won't update until the next react side affect trigger.  So the concurrent call to render will trigger multiple call because the variable is not update right away:
Screenshot from 2023-02-28 13-43-55
For the page refresh, the fetchData got called 3x instead of one time.  This is because of the data table component trigger changes when it initializing that cause a few handle to fire at the same time during the same cycle of the same state.

eg. 
1. Page render, lastFetch = '', because React.useState('')
    - 1a. fetchData got trigger by useAffect, -> setLastFetch only update the state.  lastFetch still have empty value
    - 1b. page update handler got triggered, fetchData got call, lastFect = '', thus trigger another remote data fetch
    - 1c. page size update handler got triggered, fetchData got call, lastFect = '', thus trigger another remote data fetch

File: src/bolt/utils.ts
Comment (hauvong) [reply src/bolt/utils.ts:247]: Good catch, yes, it's a typo.

File: client/src/lib/RenderReviewList.js
Comment (hauvong) [reply client/src/lib/RenderReviewList.js:153]: It's not possible because the data is coming from react-select component build from channelOptions.  But for good practice, yes, it's better to assume otherwise.  Thanks.

File: client/src/lib/RenderReviewList.js
Comment (pattiereaves) [reply client/src/lib/RenderReviewList.js]: I see what you're saying. This still feels like the wrong way for doing this. Could you check `loading === false` at the top of `fetchData` and only let it run if it's not currently executing a fetch? 

I think a more React way of doing this would be to do

```
// current channel, limit, page, status, description are all state variables

// You have handle functions that do this
const handlePageChange = (page) => {
  setCurrentPage(page)
}

const handlePerRowsChange = (newPerPage, page) => {
  setCurrentPage(page);
  setPerPage(newPerPage);
}

// and etc. with the other handle changes, then
React.useEffect(() => {
  fetchData({ channel, limit, page, status, description }); // these don't even need to be arguments any more, since they are in state, fetchData would know them
}, [channel, limit, page, status, description]);
```
then any time one of `channel, limit, page, status, description` changes then that useEffect will run the fetch.

Comment (pattiereaves) [reply client/src/lib/RenderReviewList.js]: I don't think this refactor is blocking merging these highly anticipated updates :D but a refactor in the near future is very much worth the time and will prevent bugs.

**Review by pattiereaves — APPROVED** [review]

File: client/src/lib/RenderReviewList.js
Comment (hauvong) [reply client/src/lib/RenderReviewList.js]: Check loading, it's the same because it reference useState.  The value return is a snapshot.  setLoading will not happen until the next cycle.  Unless we move the fetchData outside of the component function and implement as global, there is no other way around the duplicate calls.  Another word, the fetchData is just a helper function, not part of a state.  The fetchData var is part of the fetchData global tracking and not part of the component state value.

**Pull request merged**
